stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -0,0 +1,937 @@
+         llvm::zip(pred, onTrue, onFalse)) {
+      result.push_back(predEl != 0 ? onTrueEl : onFalseEl);
+    }
+
+    rewriter.replaceOpWithNewOp<ConstantOp>(
+        op, DenseIntElementsAttr::get(op.getResult().getType(), result));
+    return success();
+  }
+};
+
+struct EvalSliceOpPattern : public OpRewritePattern<SliceOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(SliceOp op,
+                                PatternRewriter& rewriter) const override {
+    auto resultType = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!resultType || resultType.getRank() != 1)
+      return rewriter.notifyMatchFailure(op, "expected 1-dimensional type");
+
+    SmallVector<int64_t> operand;
+    if (failed(matchInts(op.getOperand(), operand)))
+      return rewriter.notifyMatchFailure(op, "expected constant operand");
+
+    int64_t start = op.getStartIndices().getValues<int64_t>()[0];
+    int64_t limit = op.getLimitIndices().getValues<int64_t>()[0];
+    int64_t stride = op.getStrides().getValues<int64_t>()[0];
+    SmallVector<int64_t> result;
+    for (auto i = start; i < limit; i += stride) {
+      result.push_back(operand[i]);
+    }
+
+    rewriter.replaceOpWithNewOp<ConstantOp>(
+        op, DenseIntElementsAttr::get(resultType, result));
+    return success();
+  }
+};
+
+struct EvalSubtractOpPattern : public OpRewritePattern<SubtractOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(SubtractOp op,
+                                PatternRewriter& rewriter) const override {
+    return evalBinary(rewriter, op,
+                      [&](APInt lhs, APInt rhs) { return lhs - rhs; });
+  }
+};
+
+// The patterns below implement shape refinement of individual ops.
+// In a nutshell, they use the upstream type inference infrastructure and a
+// StableHLO-specific extension to refine return types based on potentially
+// refined operands.
+
+// Refines the values using the given types.
+// Tricky implementation details:
+//   1) Need to support partial shape refinements, e.g. if just a single
+//      dimension size out of an entire tensor type got refined. This is done
+//      via inferMostSpecificType.
+//   2) Need to signal propagation of the refined shapes across the
+//      StableHLO program. Different callers of this function have different
+//      propagation needs, so this function doesn't signal anything on its own
+//      and leaves that to the callers.
+LogicalResult refineValues(PatternRewriter& rewriter, Operation* op,
+                           ValueRange values, TypeRange types) {
+  if (values.size() != types.size())
+    return rewriter.notifyMatchFailure(op, [&](Diagnostic& diag) {
+      diag << "refineValues failed for " << types << ": expected "
+           << values.size() << " types, got " << types.size();
+    });
+
+  // Check whether `maybeRefinedReturnTypes` contain any new information with
```

Comment:
what is `maybeRefinedReturnTypes`, maybe an old var name of `TypeRange types`?

Reply:
Good catch! Updated the comment.